pexconn.go

```diff
@@ -0,0 +1,116 @@
+package torrent
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/anacrolix/log"
+
+	pp "github.com/anacrolix/torrent/peer_protocol"
+)
+
+const (
+	pexRetryDelay = 10 * time.Second
+	pexInterval   = 1 * time.Minute
+)
+
+// per-connection PEX state
+type pexConnState struct {
+	enabled bool
+	xid     pp.ExtensionNumber
+	seq     int
+	timer   *time.Timer
+	gate    chan struct{}
+	readyfn func()
+	torrent *Torrent
+	Listed  bool
+	info    log.Logger
+	dbg     log.Logger
+}
+
+func (s *pexConnState) IsEnabled() bool {
+	return s.enabled
+}
+
+// Init is called from the reader goroutine upon the extended handshake completion
+func (s *pexConnState) Init(c *PeerConn) {
+	xid, ok := c.PeerExtensionIDs[pp.ExtensionNamePex]
+	if !ok || xid == 0 || c.t.cl.config.DisablePEX {
+		return
+	}
+	s.xid = xid
+	s.seq = 0
+	s.torrent = c.t
+	s.info = c.t.cl.logger
+	s.dbg = c.logger
+	s.readyfn = c.tickleWriter
+	s.gate = make(chan struct{}, 1)
+	s.timer = time.AfterFunc(0, func() {
+		s.gate <- struct{}{}
+		s.readyfn() // wake up the writer
+	})
+	s.enabled = true
+}
+
+// schedule next PEX message
+func (s *pexConnState) sched(delay time.Duration) {
+	s.timer.Reset(delay)
+}
+
+// generate next PEX message for the peer; returns nil if nothing yet to send
+func (s *pexConnState) genmsg() *pp.PexMsg {
+	tx, seq := s.torrent.pex.Genmsg(s.seq)
+	if tx.Len() == 0 {
+		return nil
+	}
+	s.seq = seq
+	return tx
+}
+
+// Share is called from the writer goroutine if when it is woken up with the write buffers empty
+func (s *pexConnState) Share(postfn messageWriter) {
+	select {
+	case <-s.gate:
+		if tx := s.genmsg(); tx != nil {
+			s.dbg.Print("sending PEX message: ", tx)
+			postfn(tx.Message(s.xid))
```

Comment:
I think you might need to return the bool here, in case there are further write attempts to the post function, even though at present sharing PEX is the last thing checked.

Reply:
It looks like uploads occur after PEX, so we do want to handle this I think.

Reply:
Good point. Let me take care of this.

Reply:
Here, #394.